File: Data Structures & Algorithms/anagram-groups/submission-0.java
class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        HashMap<String, List<String>> map = new HashMap<>();
        for (String string : strs) {
            int[] array = getArray(string);
            String key = Arrays.toString(array);
            List<String> currentList = map.get(key);
            if (currentList == null) {
                currentList = new ArrayList<>();
            }
            currentList.add(string);
            map.put(key, currentList);
        }
        List<List<String>> answer = new ArrayList<>();
        for (String key : map.keySet()){
            answer.add(map.get(key));
        }
        return answer;
    }

    private int[] getArray(String string) {
        int[] array = new int[26];
        for (int index = 0; index < string.length(); index++) {
            char character = string.charAt(index);
            array[character - 'a']++;
        }
        return array;
    }
    private void print(int[] array) {
        for (int i : array) {
            System.out.print(i);
        }
        System.out.println();
    }
}

File: Data Structures & Algorithms/coin-change-ii/submission-0.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        dp = {}
        def helper(amount, index):
            if amount < 0 or index >= len(coins):
                return 0
            if (amount, index) in dp:
                return dp[amount, index]
            if amount == 0:
                return 1
            answer = helper(amount - coins[index], index) + helper(amount, index + 1)
            dp[(amount, index)] = answer
            return answer

        return helper(amount, 0)        
